R/server.R

```diff
@@ -802,21 +830,41 @@ runExample <- function(example=NA,
 #'
 #' @export
 runGadget <- function(app, server = NULL, port = getOption("shiny.port"),
-  viewer = paneViewer()) {
+  viewer = paneViewer(), stopOnCancel = TRUE) {
 
   if (!is.shiny.appobj(app)) {
     app <- shinyApp(app, server)
   }
 
+  if (isTRUE(handleCancel)) {
+    app <- decorateServerFunc(app, function(input, output, session) {
+      observeEvent(input$cancel, {
+        stopApp(stop("User cancel"))
+      })
+    })
+  }
+
   if (is.null(viewer)) {
     viewer <- browseURL
   }
 
-  retVal <- withVisible(shiny::runApp(app, port = port, launch.browser = viewer))
-  if (retVal$visible)
-    retVal$value
-  else
-    invisible(retVal$value)
+  shiny::runApp(app, port = port, launch.browser = viewer)
+}
+
+# Add custom functionality to a Shiny app object's server func
+decorateServerFunc <- function(appobj, serverFunc) {
+  origServerFuncSource <- appobj$serverFuncSource
+  appobj$serverFuncSource <- function() {
+    origServerFunc <- origServerFuncSource()
+    function(input, output, session) {
+      serverFunc(input, output, session)
+      if (length(formals(plot)) == 2)
```

Comment:
Ugh. I must be a little rusty after the holidays ;)

I'm not sure what you mean by "already been checked". We're replacing `appobj$serverFuncSource` which is the function that returns the user's server function, so Shiny will check against the function that we return (it'll always have input, output, session).

It's origServerFunc that needs to be checked, so I'm doing that here (well, I would be if I hadn't copied-and-pasted `plot` instead of `origServerFunc`). The check that is normally performed is more complicated because it checks for both clientData and session, I haven't bothered with that here but I certainly could.